### PR TITLE
HRINT-3430 & RavenDB-21788 & RavenDB-21787 Test getting certificate from Pebble

### PR DESCRIPTION
**Pull request by hiddenshadow21**
### Issue link

https://issues.hibernatingrhinos.com/issue/HRINT-3430

### Additional description

Added test that uses Pebble for acquiring certificate.

_Please delete below the options that are not relevant_

### Type of change

- Bug fix
- Regression bug fix
- Optimization
- New feature

### How risky is the change?

- Moderate 

### Backward compatibility

- Ensured. Please explain how has it been implemented?
Added default value for acme directory path that was hardcoded before.

### Is it platform specific issue?

- No

### Documentation update

- No documentation update is needed 

### Testing by Contributor

- Tests have been added that prove the fix is effective or that the feature works
- It has been verified by manual testing

### Testing by RavenDB QA team

- No special testing by RavenDB QA team is needed

### Is there any existing behavior change of other features due to this change?

- No

### UI work

- No UI work is needed

### CODE REVIEW

File: src/Raven.Server/Commercial/LetsEncryptClient.cs [hunk 1]
```diff
@@ -73,9 +74,10 @@ private static RavenHttpClient GetCachedClient(string url)
         private List<AuthorizationChallenge> _challenges = new List<AuthorizationChallenge>();
         private Order _currentOrder;
 
-        public LetsEncryptClient(string url)
+        public LetsEncryptClient(string url, string directoryPath)
```
Comment (gregolsky) [inline src/Raven.Server/Commercial/LetsEncryptClient.cs]:
```suggestion
        public LetsEncryptClient(string url) : base(url, "directory") {}

        public LetsEncryptClient(string url, string directoryPath)
```

File: src/Raven.Server/Commercial/SetupWizard/LetsEncryptSetupUtils.cs [hunk 2]
```diff
@@ -21,8 +22,8 @@ public static async Task<byte[]> Setup(SetupInfo setupInfo,  SetupProgressAndRes
             
             if (EmailValidator.IsValid(setupInfo.Email) == false)
                 throw new ArgumentException("Invalid e-mail format" + setupInfo.Email);
-
-            var acmeClient = new LetsEncryptClient(useProduction ? ProductionAcmeClientUrl : StagingAcmeClientUrl);
+            
+            var acmeClient = new LetsEncryptClient(useProduction ? ProductionAcmeClientUrl : StagingAcmeClientUrl, DirectoryPath);
```
Comment (gregolsky) [inline src/Raven.Server/Commercial/SetupWizard/LetsEncryptSetupUtils.cs]: These should probably come from configuration.

**Review by gregolsky — APPROVED** [review]

File: src/Raven.Server/Config/Categories/CoreConfiguration.cs [hunk 3]
```diff
@@ -76,10 +76,15 @@ public class CoreConfiguration : ConfigurationCategory
         public string SetupResultingServerCertificatePath { get; set; }
 
         [Description("The URLs which the server should contact when requesting certificates with the ACME protocol.")]
-        [DefaultValue("https://acme-v02.api.letsencrypt.org/directory")]
+        [DefaultValue("https://acme-v02.api.letsencrypt.org")]
```
Comment (ppekrol) [inline src/Raven.Server/Config/Categories/CoreConfiguration.cs]: this is a breaking change, you need to be able to extract that from the url IMO

**Comment by ppekrol** [comment]
Check failing tests, at least 2 are related

**Comment by ppekrol** [comment]
Please rebase and check Pebble test

File: test/SlowTests/Authentication/AuthenticationLetsEncryptTests.cs [hunk 4]
```diff
@@ -61,6 +63,23 @@ public async Task CanGetLetsEncryptCertificateAndRenewIt()
             await RenewCertificate(serverCert, firstServerCertThumbprint);
         }
 
+        private static async Task RemoveAcmeCache(string acmeUrl)
+        {
+            var home = Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData,
+                Environment.SpecialFolderOption.Create);
+            var caches = Directory.GetFiles(home, "*.lets-encrypt.cache.json");
```
Comment (ppekrol) [inline test/SlowTests/Authentication/AuthenticationLetsEncryptTests.cs]: don't you need just to do this:
```
            var home = Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData,
                Environment.SpecialFolderOption.Create);

            var hash = SHA256.Create().ComputeHash(Encoding.UTF8.GetBytes(url));
            var file = Jws.Base64UrlEncoded(hash) + ".lets-encrypt.cache.json";
```

Would expect some internal static method in the client to calculate this used in LetsEncryptClient ctor and here.
Then you do not need to actually read the file, just remove it?

File: test/SlowTests/Authentication/AuthenticationLetsEncryptTests.cs [hunk 5]
```diff
@@ -61,6 +63,23 @@ public async Task CanGetLetsEncryptCertificateAndRenewIt()
             await RenewCertificate(serverCert, firstServerCertThumbprint);
         }
 
+        private static async Task RemoveAcmeCache(string acmeUrl)
```
Comment (ppekrol) [inline test/SlowTests/Authentication/AuthenticationLetsEncryptTests.cs]: RemoveAcmeCacheAsync

File: test/SlowTests/Authentication/AuthenticationLetsEncryptTests.cs [hunk 6]
```diff
@@ -61,6 +63,23 @@ public async Task CanGetLetsEncryptCertificateAndRenewIt()
...
+                Environment.SpecialFolderOption.Create);
+            var caches = Directory.GetFiles(home, "*.lets-encrypt.cache.json");
+            foreach (string cache in caches)
+            {
+                var json = JObject.Parse(await File.ReadAllTextAsync(cache));
+                var cacheBaseUrl = new Uri(json["Location"].Value<string>()).GetLeftPart(UriPartial.Authority);
+                var acmeBaseUrl = new Uri(acmeUrl).GetLeftPart(UriPartial.Authority);
+                if (string.CompareOrdinal(cacheBaseUrl, acmeBaseUrl) != 0)
+                    continue;
+                File.Delete(cache);
```
Comment (ppekrol) [inline test/SlowTests/Authentication/AuthenticationLetsEncryptTests.cs]: IOExtensions.DeleteFile

File: test/SlowTests/Authentication/AuthenticationLetsEncryptTests.cs (hunk 5 above)
Comment (hiddenshadow21) [reply test/SlowTests/Authentication/AuthenticationLetsEncryptTests.cs]: No longer an async call.

**Review by ppekrol — APPROVED** [review]

**Pull request merged**